thriftpool/channel.go

```diff
@@ -0,0 +1,120 @@
+package thriftpool
+
+import (
+	"sync/atomic"
+)
+
+type channelPool struct {
+	pool           chan Client
+	opener         ClientOpener
+	numActive      int32
+	minConnections int
+	maxConnections int
+}
+
+// Make sure channelPool implements Pool interface.
+var _ Pool = (*channelPool)(nil)
+
+// NewChannelPool creates a new thrift client pool implemented via channel.
+func NewChannelPool(minConnections, maxConnections int, opener ClientOpener) (Pool, error) {
+	if minConnections > maxConnections {
+		return nil, &ConfigError{
+			MinConnections: minConnections,
+			MaxConnections: maxConnections,
+		}
+	}
+
+	pool := make(chan Client, maxConnections)
+	for i := 0; i < minConnections; i++ {
+		c, err := opener()
+		if err != nil {
+			return nil, err
+		}
+		pool <- c
+	}
+
+	return &channelPool{
+		pool:           pool,
+		opener:         opener,
+		minConnections: minConnections,
+		maxConnections: maxConnections,
+	}, nil
+}
+
+// Get returns a thrift client from the pool.
+func (cp *channelPool) Get() (client Client, err error) {
+	defer func() {
+		if err == nil {
+			atomic.AddInt32(&cp.numActive, 1)
+		}
+	}()
+
+	select {
+	case c := <-cp.pool:
+		if c.IsOpen() {
+			return c, nil
+		}
+	default:
```

Comment:
Am I correct in understanding that this means that if the pool is exhausted, we don't block at all to wait for a connection to be released?  That seems like something we might want to support rather than immediately returning an error?

Reply:
Ah, I forgot that we don’t necessarily start with a full pool, this makes more sense given that.

Reply:
Yes :)